### PR TITLE
[TensorExpr] Move AOT compilation logic from aot_compiler.cpp to NNC's to_backend

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70475
* __->__ #70375
* #70374

Differential Revision: [D33303645](https://our.internmc.facebook.com/intern/diff/D33303645)

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 1]
```diff
@@ -189,6 +194,153 @@ std::pair<std::unique_ptr<Function>, const std::string> aotCompile(
...
+std::shared_ptr<Graph> preprocessGraphPasses(
+    std::shared_ptr<Graph>& graph,
+    const std::vector<c10::optional<at::Tensor>>& example_inputs) {
+  torch::jit::RemoveTensorMutation(graph);
+  torch::jit::EliminateDeadCode(graph->block());
+  graph = torch::jit::tensorexpr::removeUnusedSelfArgument(graph);
+
+  torch::jit::tensorexpr::annotateInputShapes(graph, example_inputs);
+  torch::jit::OptimizeFrozenGraph(graph, true);
+  torch::jit::PropagateShapesOnGraph(graph);
```
Comment (priyaramani) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:281]: Do you think we can remove the shape propagation passes as TraceGraph() is called from aotCompile() ?

File: binaries/aot_model_compiler.cc [hunk 2]
```diff
@@ -217,27 +112,6 @@ int main(int argc, char** argv) {
...
-  auto graph = frozen_m.get_method(FLAGS_method_name).graph();
-  auto inputShapes = parseInputShapes();
-  auto inputTypes = parseInputTypes();
-  std::vector<c10::optional<at::Tensor>> example_inputs;
-  example_inputs.reserve(inputShapes.size());
-  for (int i = 0; i < inputShapes.size(); ++i) {
-    example_inputs.emplace_back(at::rand(inputShapes[i]).to(at::dtype(inputTypes[i])));
-  }
-
-  torch::jit::RemoveTensorMutation(graph);
```
Comment (priyaramani) [inline binaries/aot_model_compiler.cc:229]: Can we also remove the corresponding headers imported for this and others?

**Review by priyaramani — APPROVED** [review]

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

File: binaries/aot_model_compiler.cc (hunk 2 above)
Comment (ZolotukhinM) [reply binaries/aot_model_compiler.cc:229]: Good point, thanks!

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp (hunk 1 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/mobile/nnc/aot_compiler.cpp:306]: I'd prefer to keep it (but I don't feel strongly about it). My motivation is that 1) it doesn't hurt to run shape-prop if all the shapes are already there, 2) in theory it's possible to invoke `preprocess` from some other place than `aotCompile` - in that case shape-prop can actually do work.

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

File: test/mobile/nnc/test_nnc_backend.cpp [hunk 3]
```diff
@@ -187,49 +95,6 @@ TEST(NNCBackendTest, AOTCompileThenExecute) {
   EXPECT_TRUE(result.toTensor().equal(reference.toTensor()));
 }
 
-TEST(NNCBackendTest, FakeTensor) {
```
Comment (ZolotukhinM) [inline test/mobile/nnc/test_nnc_backend.cpp:190]: I had to remove this test as we won't be able to compile models with calls to methods of custom classes (at least as of now). The other test still exercises the AOT workflow.

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

File: binaries/aot_model_compiler.cc [hunk 4]
```diff
@@ -61,125 +54,20 @@ std::vector<std::string> split(
...
   c10::Dict<c10::IValue, c10::IValue> method_spec(
       c10::StringType::get(), c10::AnyType::get());
-  auto inputShapes = parseInputShapes();
-  auto inputTypes = parseInputTypes();
-  method_spec.insert("sizes", inputShapes);
-  method_spec.insert("types", inputTypes);
+  method_spec.insert("sizes", FLAGS_input_dims);
+  method_spec.insert("types", FLAGS_input_types);
+  method_spec.insert("asmfile", FLAGS_output_llvm);
+  method_spec.insert("model_name", FLAGS_model_name);
```
Comment (priyaramani) [inline binaries/aot_model_compiler.cc:65]: model_name and model_version can be in the compile_spec outside the method_spec ?

File: torch/csrc/jit/mobile/nnc/aot_compiler.cpp [hunk 5]
```diff
@@ -178,40 +183,184 @@ std::pair<std::unique_ptr<Function>, const std::string> aotCompile(
...
+  for (auto example_input : example_inputs) {
+    example_values.emplace_back(*example_input);
   }
+  graph = TraceGraph(graph, example_values);
+  tensorexpr::annotateInputShapes(graph, example_inputs);
 
-  GRAPH_DUMP("graph before compiler passes ", g);
-  tensorexpr::removeUnusedSelfArgument(g);
-  g = TraceGraph(g, example_values);
   // TODO: Remove annotateInputShapes pass when TraceGraph can also capture
```
Comment (priyaramani) [inline torch/csrc/jit/mobile/nnc/aot_compiler.cpp:300]: ultra nit: Have the comment with annotateInputShapes removing the empty line in between

**Review by priyaramani — APPROVED** [review]

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

File: binaries/aot_model_compiler.cc (hunk 4 above)
Comment (ZolotukhinM) [reply binaries/aot_model_compiler.cc:65]: There is a bunch of places where we iterate over compile spec and expect keys to be method names, so we can't just put model_name and model_version there. Would you mind if I clean that up in a separate PR after this lands?

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33303645).

File: binaries/aot_model_compiler.cc (hunk 4 above)
Comment (priyaramani) [reply binaries/aot_model_compiler.cc:65]: Ah that's right there are places where we expect keys to be method names. I'll look at what other delegates are doing and converge them.

**Pull request closed without merging**